flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlStateTestBase.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state.ttl;
+
+import org.apache.flink.api.common.time.Time;
+import org.apache.flink.util.function.SupplierWithException;
+import org.apache.flink.util.function.ThrowingConsumer;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+abstract class TtlStateTestBase<S, UV, GV> {
```

Comment:
I think it would make sense to extend this general test a bit to consider multiple keys and namespaces. Ideally, a test should really test the full contract specification of the tested subject. What is mean is, you could currently pass this test even if TTL would accidentally clear all states on the timeout of one state, or maybe clear all the states in the same namespace. The mock states can easily be extended to truly scope values by key and namespace. Then the test can, for example, create two keys in the same namespace and two keys in a different namespace and check that their timeouts are isolated from each other and the interaction works as expected,

Reply:
Good point, TTL states are implemented on top of interfaces which do not operate on keys, but I think it should be part of final integration tests with backends which depend on keyed context.